README.md
# JiraFirebase

## Functionality

This code reads issues from JIRA, then stores them in FireBase. It then checks the last stored version, and compares it to the newly added version. Any differences are noted, then sent as a mass email. The last stored version is deleted, and the new version replaces it.
This code is designed to be run every few minutes.

## Backstory

This was created as part of an internship I did with a game development company over the summer of 2017. The company I was working with used SVN for version control. As such, this repo doesn't have many commits outside of some setup to put it on GitHub. This is not because I wasn't using version control, but because I wasn't using GitHub for it.
This is also why I don't have a .gitignore, since I did all that cleanup with SVN.

## Downloading

If you were to download the code now, it would not run. Why? Because the functional version contains sensitive information for both the company I was interning at and its employees (see [the backstory of this repo](#backstory)). So I do not recommend downloading the code to use it. However, if you are looking to do something similar, feel free to look through the code and [shoot me an email](mailto:hazel.r.pearson@gmail.com) if you have any questions!

## Licensing

This code is licensed under the GNU General Public License v3.0. To summarize, the license...

GNU GPLv3 means you can:
- Use my code commercially
- Distribute it
- Modify it
- Patent it
- Use it privately

But have to:
- Disclose its source (here!)
- Provide the [copyright notice](#legalese) and license, which appears in my LICENSE.md
- License your code under this same license (GNU GPLv3)
- Document what changes you make to my code

And:
- You don't have a warranty
- I'm not liable

Copyright (C) 2017 Hazel Pearson

### Legalese

This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program.  If not, see [http://www.gnu.org/licenses/](http://www.gnu.org/licenses/).

You can contact me with any questions about licensing [here](mailto:hazel.r.pearson@gmail.com).

changes.txt
==== Scanning at Thu Aug 24 16:23:38 PDT 2017 ====
==== Scanning at Sat Sep 16 10:27:35 PDT 2017 ====
Resolved issue FLO-3218 (scott@flowplay.com): "HX: Charm Parties: Error #3681 Texture size is zero "

Resolved issue FLO-3266 (kristen@flowplay.com): "VG: Guy's Copacetic Set, var2 items are misshapen"

Changed issue FLO-3202 (colin@flowplay.com): "HX: Non-fatal exception thrown occasionally when doing dance moves."
Changes:
	Last Updated | Fri Sep 15 15:35:00 PDT 2017 -> Sat Sep 16 10:09:03 PDT 2017
	Comments | {parentIssueType=Bug, author=scottc@flowplay.com, id=16233, body=<p>I'm getting this 100% using the first dance (Icon shown in screenshot). </p>

<p>1. Login as a male avatar (guy6@vegasworld.com / pass is what I was using)<br/>
2. Navigate to social room<br/>
3. Attempt to use the first dance on the list</p>

<p>Actual Result<br/>
The exception occurs, and the button no longer functions. </p>

<p>Expected Result<br/>
Avatar is able get down with all available dance moves. </p>, parentBugId=FLO-3202} -> {parentIssueType=Bug, author=scottc@flowplay.com, id=16233, body=<p>I'm getting this 100% using the first dance (Icon shown in screenshot). </p>

<p>1. Login as a male avatar (guy6@vegasworld.com / pass is what I was using)<br/>
2. Navigate to social room<br/>
3. Attempt to use the first dance on the list</p>

<p>Actual Result<br/>
The exception occurs, and the button no longer functions. </p>

<p>Expected Result<br/>
Avatar is able get down with all available dance moves. </p>, parentBugId=FLO-3202}, {parentIssueType=Bug, author=chance@flowplay.com, id=16313, body=<p>Reopening. This still happens with various dance moves. Using a female avatar, you can see it at least with some of the dance moves in the bottom three rows of moves... once you get through the numerous warning messages which I think are due to 
    <span class="jira-issue-macro" data-jira-key="FLO-2371" >
                <!-- replace the span with an AUI template -->
        <a href="https://flowplay.atlassian.net/browse/FLO-2371" class="jira-issue-macro-key issue-link"  title="FX: Create ddr Dance Move icons" >
            <img class="icon" src="https://flowplay.atlassian.net/secure/viewavatar?size=xsmall&avatarId=10303&avatarType=issuetype" />
            FLO-2371
        </a>
                                                    <span class="aui-lozenge aui-lozenge-subtle aui-lozenge-complete jira-macro-single-issue-export-pdf">NEW</span>
            </span>
.</p>

<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>RangeError: Error #1125: The index 64 is out of range 64.
	at avatar.dynamic_bitmap::AnimDefinition/readFromBin()[SharedX/Avatar/src/avatar/dynamic_bitmap/AnimDefinition.hx:133]
	at avatar.dynamic_bitmap::AnimDefinition()[SharedX/Avatar/src/avatar/dynamic_bitmap/AnimDefinition.hx:50]
	at avatar.dynamic_bitmap::Animation3D/handleCompleteBin()[SharedX/Avatar/src/avatar/dynamic_bitmap/Animation3D.hx:534]
	at flash.events::EventDispatcher/dispatchEventFunction()
	at flash.events::EventDispatcher/dispatchEvent()
	at flash.net::URLLoader/onComplete()
</pre>
</div></div>, parentBugId=FLO-3202}, {parentIssueType=Bug, author=colin@flowplay.com, id=16316, body=<p>animation data is stored in a collection of fixed length vectors, so I've been bumping up size to accommodate longer anims. set to 100 now.</p>, parentBugId=FLO-3202}

Resolved issue FLO-3255 (kristen@flowplay.com): "VG: Girl's Oh My Goth Shoes, red color on var1 is too bright"

Resolved issue FLO-3249 (devin@flowplay.com): "VG: Girl's Spooky Bride Gown should be in Dresses category"

Resolved issue FLO-3258 (colin@flowplay.com): "HX: Incentive appearing during guest registering"

Resolved issue FLO-3221 (david@flowplay.com): "HX:  Avatars: Charm Parties: Uncaught exception: Error #1009: Cannot access a property or method of a null object reference occurs"

==== Scanning at Sat Sep 16 10:34:17 PDT 2017 ====
Resolved issue FLO-3218 (scott@flowplay.com): "HX: Charm Parties: Error #3681 Texture size is zero "

Resolved issue FLO-3266 (kristen@flowplay.com): "VG: Guy's Copacetic Set, var2 items are misshapen"

Changed issue FLO-3202 (colin@flowplay.com): "HX: Non-fatal exception thrown occasionally when doing dance moves."
Changes:
	Last Updated | Fri Sep 15 15:35:00 PDT 2017 -> Sat Sep 16 10:09:03 PDT 2017
	Comments | {parentIssueType=Bug, author=scottc@flowplay.com, id=16233, body=<p>I'm getting this 100% using the first dance (Icon shown in screenshot). </p>

<p>1. Login as a male avatar (guy6@vegasworld.com / pass is what I was using)<br/>
2. Navigate to social room<br/>
3. Attempt to use the first dance on the list</p>

<p>Actual Result<br/>
The exception occurs, and the button no longer functions. </p>

<p>Expected Result<br/>
Avatar is able get down with all available dance moves. </p>, parentBugId=FLO-3202} -> {parentIssueType=Bug, author=scottc@flowplay.com, id=16233, body=<p>I'm getting this 100% using the first dance (Icon shown in screenshot). </p>

<p>1. Login as a male avatar (guy6@vegasworld.com / pass is what I was using)<br/>
2. Navigate to social room<br/>
3. Attempt to use the first dance on the list</p>

<p>Actual Result<br/>
The exception occurs, and the button no longer functions. </p>

<p>Expected Result<br/>
Avatar is able get down with all available dance moves. </p>, parentBugId=FLO-3202}, {parentIssueType=Bug, author=chance@flowplay.com, id=16313, body=<p>Reopening. This still happens with various dance moves. Using a female avatar, you can see it at least with some of the dance moves in the bottom three rows of moves... once you get through the numerous warning messages which I think are due to 
    <span class="jira-issue-macro" data-jira-key="FLO-2371" >
                <!-- replace the span with an AUI template -->
        <a href="https://flowplay.atlassian.net/browse/FLO-2371" class="jira-issue-macro-key issue-link"  title="FX: Create ddr Dance Move icons" >

src/individuals/IssueJira.java
package individuals;

import general.Fields;
import general.Fields.Field;
import general.JsonParser;
import general.MainJira;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeFormatterBuilder;
import java.util.*;

import static general.Fields.Field.id;
import static general.MainJira.getDataFromURL;
import static java.time.format.DateTimeFormatter.ISO_LOCAL_DATE_TIME;

/*********************************************************
 * Description:
 *
 * Represents an issue in JIRA
 *
 * User: Doug
 * Date: 6/11/2017
 * Copyright FlowPlay, Inc (www.flowplay.com)
 *********************************************************/

public class IssueJira {
    private final JSONObject m_totalJson ;
    private final JSONObject m_fieldsJson ;

    private HashMap<Field, String> fieldsMap = new HashMap<>();

    public IssueJira(JSONObject json) throws JSONException {
        m_totalJson = json ;
        m_fieldsJson = m_totalJson.getJSONObject("fields") ;
        fieldsMap.put(Fields.Field.summary, JsonParser.safeGet(m_fieldsJson,"summary"));
        fieldsMap.put(Fields.Field.priority, JsonParser.safeChild(m_fieldsJson,"priority","name"));
        fieldsMap.put(Fields.Field.assignee, JsonParser.safeChild(m_fieldsJson,"assignee","emailAddress"));
        fieldsMap.put(Fields.Field.status, JsonParser.safeChild(m_fieldsJson,"status", "name"));
        fieldsMap.put(Fields.Field.creationDate, Fields.humanReadable(JsonParser.safeGet(m_fieldsJson,"created")));
        fieldsMap.put(Fields.Field.reporter, JsonParser.safeChild(m_fieldsJson,"reporter","emailAddress"));
        fieldsMap.put(Fields.Field.lastUpdated, Fields.humanReadable(JsonParser.safeGet(m_fieldsJson, "updated")));
        fieldsMap.put(Fields.Field.severity, getCustomField("customfield_10204"));
        fieldsMap.put(Fields.Field.hardware, getCustomField("customfield_10205"));
        fieldsMap.put(Fields.Field.os, getCustomField("customfield_10206"));
        fieldsMap.put(Fields.Field.patchIntoPlaytest, getCustomField("customfield_10603"));
        fieldsMap.put(Fields.Field.html5Bug, getCustomField("customfield_10802"));
        fieldsMap.put(Fields.Field.hasBeenPatched, getCustomField("customfield_10604"));
        fieldsMap.put(Fields.Field.fixBeforeHtml5, getCustomField("customfield_10607"));
        fieldsMap.put(Fields.Field.product, getCustomField("customfield_10700"));
        fieldsMap.put(Fields.Field.epicLink, JsonParser.safeGet(m_fieldsJson,"customfield_10005"));
        fieldsMap.put(Fields.Field.world, getCustomField("customfield_10602"));
        fieldsMap.put(Fields.Field.sprint, getSprint());
        fieldsMap.put(Fields.Field.sprints, getSprints());
        fieldsMap.put(id, IssueJira.getID(m_fieldsJson));
        fieldsMap.put(Fields.Field.url, "https://flowplay.atlassian.net/browse/" + fieldsMap.get(id));
        fieldsMap.put(Fields.Field.days, JsonParser.safeGet(m_fieldsJson,"customfield_10117"));
        fieldsMap.put(Field.issueType, JsonParser.safeChild(m_fieldsJson, "issuetype","name"));
        fieldsMap.put(Field.description, JsonParser.safeChild(m_totalJson, "renderedFields","description"));
        fieldsMap.put(Fields.Field.productOwner, JsonParser.safeChild(m_fieldsJson, "customfield_10900", "emailAddress"));
    }

    public HashMap<Field, String> getFieldsMap() { return fieldsMap; }

    private static Date convertJIRADate(String dateStr) {
        // Dates from JIRA look like this:
        // String date = "2017-09-26T20:14:57.473-0700"

        // ISO standard dates look like this:
        // String date = "2017-09-26T20:14:57.473-07:00"     <== 07:00 note the :

        // That causes lots of problems so we need a special parser...
        DateTimeFormatter jiraDateFormat = new DateTimeFormatterBuilder()
                .parseCaseInsensitive()
                .append(ISO_LOCAL_DATE_TIME)
                .appendOffset("+HHmm", "+00:00")
                .toFormatter();

        Date date = Date.from( ZonedDateTime.parse( dateStr , jiraDateFormat ).toInstant() ) ;
        return date ;
    }

    // These don't seem to come in a reliable order
    // so we'll search through the list to find the one that is most recent
    private static JSONObject findMostRecentChange(JSONArray changelog) throws JSONException {
        JSONObject lastChange = null ;
        Date lastDate = null ;

        for (int i = 0 ; i < changelog.length() ; i++) {
            JSONObject nextChange = changelog.getJSONObject(i);

            String createdDateStr = nextChange.getString("created");
            Date createdDate = convertJIRADate(createdDateStr);

            if (lastChange == null || lastDate == null || createdDate.after(lastDate)) {
                lastChange = nextChange ;
                lastDate = createdDate ;
            }
        }

        return lastChange ;
    }

    public static String getChangedBy(String id, String reporter, Map<String, String> lastComment) {
        String changedBy = "";
        try {
            JSONObject data = new JSONObject(getDataFromURL("https://flowplay.atlassian.net/rest/api/2/issue/" + id + "?expand=changelog"));
            JSONArray changelog = data.getJSONObject("changelog").getJSONArray("histories");
            if (changelog.length() != 0) {
                JSONObject lastChange = findMostRecentChange(changelog) ;

                String createdDateStr = lastChange.getString("created") ;
                Date createdDate = convertJIRADate(createdDateStr) ;

                changedBy = lastChange.getJSONObject("author").getString("emailAddress");

                // Comments unfortunately do not show up on the change list
                // so we need to consider who left the last comment and if that
                // was later, then that was the last person to change this bug
                if (lastComment != null) {
                    String lastCommentCreatedDateStr = lastComment.get("created") ;
                    String commentAuthor = lastComment.get("author") ;
                    Date lastCommentDate = convertJIRADate(lastCommentCreatedDateStr) ;

                    if (lastCommentDate.after(createdDate)) {
                        System.out.println("Last comment for " + id + " was " + lastCommentDate + " which is after last change " + createdDate + " so using last commenter " + commentAuthor + " not " + changedBy) ;
                        changedBy = commentAuthor ;
                    } else {
                        System.out.println("Last comment for " + id + " was not after last change so using last changer " + changedBy) ;
                    }
                } else {
                    System.out.println("No last comment for " + id + " so last changed " + changedBy) ;
                }

            } else { //No one's changed this issue, it's only been created
                if (lastComment != null) {
                    String commentAuthor = lastComment.get("author") ;
                    changedBy = commentAuthor ;
                    System.out.println("No changes for " + id + " and last commenter was " + commentAuthor);
                } else {
                    System.out.println("No changes for " + id + " so using reporter " + reporter);
                    changedBy = reporter;
                }
            }
        } catch (JSONException ex) {
            System.err.println("Error trying to retrieve who last changed issue " + id);
            ex.printStackTrace();
        }

        System.out.println("Got changed by for " + id + " " + changedBy) ;

        return changedBy;
    }

    public ArrayList<String> getWatchers() {
        return IssueJira.getWatchers(fieldsMap.get(id));
    }

    public static ArrayList<String> getWatchers(String id) {
        ArrayList<String> watchersList = new ArrayList<>();
        try {
            String jsonWatchers = MainJira.getDataFromURL("https://flowplay.atlassian.net/rest/api/2/issue/" + id + "/watchers") ;
            
            JSONObject data = new JSONObject(jsonWatchers);
            JSONArray watchers = data.getJSONArray("watchers");
            for (int i = 0; i < watchers.length(); i++) {
                JSONObject curr = watchers.getJSONObject(i);
                watchersList.add(JsonParser.safeGet(curr, "emailAddress"));
            }
        } catch (Exception ex) {
            System.err.println("Error trying to retrieve watchers for " + id);
            ex.printStackTrace();
            throw new IllegalStateException(ex) ;
        }

        System.out.println("Got watchers for " + id + " " + watchersList) ;

        return watchersList;
    }

    public ArrayList<HashMap<String, String>> getComments() {
        return IssueJira.getComments(m_totalJson, fieldsMap.get(id), fieldsMap.get(Field.issueType));
    }

    public static ArrayList<HashMap<String, String>> getComments(JSONObject totalJson, String bugId, String issueType) {
        ArrayList<HashMap<String, String>> commentList = new ArrayList<>();
        try {
            // There seem to be two representation of comments
            // one more human readable and one more machine readable
            // Both are potentially useful to us
            JSONArray rawComments = totalJson.getJSONObject("fields").getJSONObject("comment").getJSONArray("comments");
            JSONArray comments = totalJson.getJSONObject("renderedFields").getJSONObject("comment").getJSONArray("comments");
            for (int i = 0; i < comments.length(); i++) {
                JSONObject currComment = comments.getJSONObject(i);
                JSONObject currRawComment = rawComments.getJSONObject(i);

                HashMap<String, String> newComment = new HashMap<>();
                newComment.put("author", JsonParser.safeChild(currComment,"author","emailAddress"));
                newComment.put("body",JsonParser.safeGet(currComment,"body"));
                String id = JsonParser.safeGet(currComment,"id");
                newComment.put("id",id);
                newComment.put("parentBugId",bugId);
                newComment.put("parentIssueType",issueType);

                String createdStr = JsonParser.safeGet(currRawComment, "created") ;
                newComment.put("created", createdStr) ;

                commentList.add(newComment);
            }

        } catch (JSONException ex) {
            System.err.println("Error trying to retrieve comments: ");
            ex.printStackTrace();
        }

        //System.out.println("Got comments for " + bugId + " " + commentList.size()) ;

        return commentList;
    }

    private List<String> getAllSprints() {
        List<String> sprints = new ArrayList<>() ;

        if (!m_fieldsJson.isNull("customfield_10115")) {
            String data = JsonParser.safeGet(m_fieldsJson, "customfield_10115");

            /* [
            "com.atlassian.greenhopper.service.sprint.Sprint@1e49e9ba[id=23,rapidViewId=2,state=CLOSED,name=6/26 Sprint,goal=,startDate=2017-06-26T18:30:40.259Z,endDate=2017-07-08T01:00:00.000Z
                    ,completeDate=2017-07-10T18:35:58.410Z,sequence=22]",
            "com.atlassian.greenhopper.service.sprint.Sprint@5242d133[id=24,rapidViewId=6,state=CLOSED,name=7/10 Sprint (24),goal=,startDate=2017-07-10T18:30:44.797Z,endDate=2017-07-22T01:00:00
            .000Z,completeDate=2017-07-24T21:08:30.287Z,sequence=23]",
            "com.atlassian.greenhopper.service.sprint.Sprint@39d21748[id=31,rapidViewId=6,state=FUTURE,name=10/16 Sprint (31),goal=,startDate=<null>,endDate=<null>,completeDate=<null>,sequence=
            30]
            */
            boolean done = false ;
            int pos = 0 ;

            // Find the ",name=" entries and pull the value following that.
            while (!done) {
                pos = data.indexOf(",name=", pos) ;

                if (pos == -1) {
                    done = true ;
                    continue;
                }

                pos += ",name=".length() ;

                int end = data.indexOf(",", pos) ;
                String sprint = data.substring(pos, end) ;
                sprints.add(sprint) ;

                pos = end ;
            }
        }

        // Sort them in order of sprint number
        sprints.sort(new SprintSorter());

        return sprints ;
    }

    private String getSprints() {
        List<String> all = getAllSprints() ;
        return all.toString() ;
    }

    private static class SprintSorter implements Comparator<String> {
        // [11/13 Sprint (33) <= find the 33
        private static int getSprintNumber(String sprint) {
            int open = sprint.indexOf("(") ;
            int close = sprint.indexOf(")") ;
            if (open == -1 || close == -1 || open > close)
                return 0 ;

            String number = sprint.substring(open+1, close) ;
            int sprintNumber = Integer.parseInt(number) ;
            return sprintNumber ;
        }

        @Override
        public int compare(String o1, String o2) {
            int sprint1 = getSprintNumber(o1) ;
            int sprint2 = getSprintNumber(o2) ;

            return Integer.compare(sprint1, sprint2) ;
        }
    }

    public String getSprint() {
        List<String> all = getAllSprints() ;

        if (all.isEmpty())
            return "" ;

        // Return the last sprint
        return all.get(all.size()-1) ;
    }

    private String getCustomField(String customField) {
        try {
            if (m_fieldsJson.has(customField) && !m_fieldsJson.isNull(customField)) {
                JSONObject data = m_fieldsJson.getJSONObject(customField);

                return data.getString("value");
            }
        } catch (JSONException ex) {
            System.err.println("Error trying to get custom field " + customField + " from " + m_fieldsJson + ":");
            ex.printStackTrace();
            return null;
        }
        return "";
    }

    public static String getID(JSONObject issue) {
        String watchers = JsonParser.safeChild(issue,"watches","self");
        //eg "https://flowplay.atlassian.net/rest/api/2/issue/FLO-2693/watchers"

        int flo = watchers.indexOf("FLO-");
        int slash = watchers.indexOf("/",flo);
        if (slash != -1) {
            return watchers.substring(flo, slash);
        } else {
            return watchers.substring(flo);
        }
    }

    @Override
    public String toString() {
        return "JiraIssue{" + "m_fieldsJson=\n" + m_fieldsJson + "\n" ;
    }
}
